README.md
***** BEGIN LICENSE BLOCK *****
Version: MPL 1.1/GPL 2.0/LGPL 2.1

The contents of this file are subject to the Mozilla Public License Version
1.1 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at
http://www.mozilla.org/MPL/

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
for the specific language governing rights and limitations under the
License.

The Original Code is the Perl XPCOM language bindings.

The Initial Developer of the Original Code is
Jumpline.com, Inc. Portions created by Jumpline.com, Inc. are
Copyright (C) 2001 Jumpline.com, Inc. All Rights Reserved.

Contributor(s):
Matt Kennedy <matt@jumpline.com> (original author)

Alternatively, the contents of this file may be used under the terms of
either the GNU General Public License Version 2 or later (the "GPL"), or
the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
in which case the provisions of the GPL or the LGPL are applicable instead
of those above. If you wish to allow use of your version of this file only
under the terms of either the GPL or the LGPL, and not to allow others to
use your version of this file under the terms of the MPL, indicate your
decision by deleting the provisions above and replace them with the notice
and other provisions required by the GPL or the LGPL. If you do not delete
the provisions above, a recipient may use your version of this file under
the terms of any one of the MPL, the GPL or the LGPL.

***** END LICENSE BLOCK *****

Perl Interface to XPCOM
-----------------------

This package allows perl to access XPCOM objects as well as allowing XPCOM 
components to be implemented in Perl. This module is still young, and rough 
around the edges, but it is fairly functional in its current state.

Building the module
-------------------

Warning! These instructions are going to be Unix-centric until such time as 
I have time to build everything on Win32 and Mac.

These instructions cover building this module as a mozilla extension. The 
first step is to open the Perl XPCOM package into mozilla's extensions 
directory. Then you must edit the allmakefiles.sh file at the top of the 
mozilla source tree to include the extension into your build. The 
allmakefiles.sh file has a section for handling extensions, I recommend 
searching for the string 'p3p' in the file and it will take you where you 
need to go. You should find yourself at a switch statement block with the 
names of the directories in the extensions dir as the cases. Add the perl 
extension like this:

        # This section of code is already in allmakefiles.sh. I put it 
        # here to give you a reference for navigating this file.
        p3p ) MAKEFILES_extensions="$MAKEFILES_extensions
            extensions/p3p/Makefile
            extensions/p3p/public/Makefile
            extensions/p3p/resources/Makefile
            extensions/p3p/resources/content/Makefile
            extensions/p3p/resources/locale/Makefile
            extensions/p3p/resources/locale/en-US/Makefile
            extensions/p3p/resources/skin/Makefile
            extensions/p3p/src/Makefile
            " ;;
        # This is the part you need to add. Cut and paste this into 
        # allmakefiles.sh
        perl ) MAKEFILES_extensions="$MAKEFILES_extensions
            extensions/perl/Makefile
            extensions/perl/XPCOM/Makefile
            extensions/perl/src/Makefile
            extensions/perl/tests/Makefile
            extensions/perl/public/Makefile
            " ;;
        # Done

Once you've modified allmakefiles.sh, add the following line to the .mozconfig 
file in the top of the mozilla source tree:

  ac_add_options --enable-extensions=default,perl

Along with any other switches and extensions you'd like to add.

Once all that is done, you can build the lizard in the usual way (you have to use GNU make).

  make -f client.mk build

or

  ./configure && make

After Mozilla has finished building, the final step is to install the Perl 
module into Perl's include path. Do this by going into the 
extensions/perl/XPCOM directory and as root run the command:

  make module_install

Using the Extension
-------------------

I don't have a good, whiz-bang test yet. For now, the contents of the test 
directory build a component that replicates the nsSample implementation from 
the XPCOM tree. You can view the HTML page there, changing it to set the 
component contract ID to the perl component (look for '@mozilla.org/sample;1' 
and replace it with '@mozilla.org/perlsample;1'). Also included in the 
tests is the obligatory Hello World![tm]. Thrilling, eh...?

Bugs
----
You betcha.
What's implemented currently has worked without trouble. The two big gotchas 
to watch out for are:

Beware of invoking methods on Perl components from the destructor of a C++ 
object. I have found this to be the source of strange, random crashes which 
I haven't been able to track down yet.

Unicode is kludged, see this warning from the perlunicode(1) manpage:

  WARNING:  As of the 5.6.1 release, the implementation of Unicode
  support in Perl is incomplete, and continues to be highly experimental.

Because of this, I've chosen to force conversion of PRUnichar strings to 
C strings, which sucks.

Don't run a perl component outside of the main thread. Perl code can dispatch 
runnables to a thread, but the perl code itself cannot run in other threads. 
This is due to general thread unsafety in Perl and in not likely to change 
anytime soon.

Don't use external Perl modules that bind to C libraries because linking 
issues abound. By default, the perl component loader initializes its own
interpreter. Since this component is dynamically loaded by mozilla, symbols 
the perl modules need to bootstrap in the interpreter aren't exported. 
The workaround for this is to link libperl against the main application 
(only feasable if you're using XPCOM in an app other than mozilla) and 
clear the LOCAL_PERL Makefile variable in extensions/perl/src/Makefile. 
This will prevent the perl component loader from initializing its own 
interpreter.

TODO
----
Currently, I still haven't added Win32 makefiles.

Array types are partially implemented.

The XPCOM perl module package contains sub modules called XPCOM::GenericModule 
and XPCOM::GenericFactory which make it much less painful to create new 
components, unfortunately I haven't had time to write the POD docs for them 
yet!

Matt Kennedy <matt@jumpline.com>
Last Modified: Wed Oct 24 15:11:16 EDT 2001



ChangeLog
2001-10-24 15:24  matt

	* tests/: catclsids.pl, nsSelfTest.pl, plPerlBool.pm: Added tests
	and examples.

2001-10-24 15:24  matt

	* src/: Makefile.in, plPerlBase.cpp, plPerlComponentLoader.cpp,
	plPerlXPTCStub.cpp: Cleanups, added plPerlComponentHelper
	implementation.

2001-10-24 15:22  matt

	* public/Makefile.in: Renamed module name because the old name was
	poor use of namespace.

2001-10-24 15:21  matt

	* doc/: xpcom-perl.html, xpcom-perl.xml: Update docs for new stuff.

2001-10-24 15:21  matt

	* XPCOM/: GenericFactory/GenericFactory.pm,
	GenericModule/GenericModule.pm: Update version number.

2001-10-24 15:21  matt

	* XPCOM/: Components.xs, Makefile.in, XPCOM.pm: Implemented
	Components::WrapObject and Components::Manager

2001-10-24 15:21  matt

	* ChangeLog, README: Added ChangeLog

2001-10-22 23:13  matt

	* src/plPerlComponentLoader.cpp: Fix for the new frozen
	nsIObserverService, which the perl component loader sends
	notification events to.
	
	Maintains back compatibility with the old API as that is what
	everyone is currently using unless their on the bleeding edge of
	the Mozilla HEAD branch.

2001-10-22 17:15  matt

	* XPCOM/Makefile.in: Minor dependency fix that was breaking builds
	in Mozilla.

2001-10-22 15:36  matt

	* INSTALL, README: Updated install stuff for added public
	directory.

2001-10-22 15:27  matt

	* Makefile.in, XPCOM/Components.xs, XPCOM/Makefile.PL,
	XPCOM/XPCOM.pm, XPCOM/nsISupports.xs, XPCOM/perlXPUtil.cxx,
	XPCOM/perlXPUtil.h, public/Makefile.in,
	public/plIPerlComponentHelper.idl, src/Makefile.in,
	src/perlXPUtil.h, src/plPerlBase.cpp,
	src/plPerlComponentLoader.cpp, src/plPerlXPCOM_Base.h,
	src/plPerlXPModule.cpp, src/plPerlXPTCStub.cpp,
	tests/nsTestPerlSample.cpp: 
	Added supports for DOMString types, so Mozilla core DOM API is
	fully accessible now (including all the XML extras stuff).
	
	Initial support for the array type (input only right now). Output
	arrays are yet to be done, plus perl components cannot implement
	array types themselves.
	
	Self referential support has been added via the
	Components::WrapObject() subroutine. This wraps a perl blessed
	object in an XPCOM binding. This allows perl component objects to
	pass pointers to themselves to other XPCOM methods.

2001-10-12 14:13  matt

	* src/plPerlBase.cpp: Fix compile time error on newer gcc.

2001-10-09 23:10  matt

	* XPCOM/Makefile.in, src/Makefile.in: Landed changes for new
	sandbox and build layout.

2001-10-09 17:29  matt

	* INSTALL, Makefile.in, README, XPCOM/Changes, XPCOM/Components.xs,
	XPCOM/MANIFEST, XPCOM/Makefile.PL, XPCOM/Makefile.in, XPCOM/README,
	XPCOM/XPCOM.pm, XPCOM/nsISupports.xs, XPCOM/perlXPUtil.cxx,
	XPCOM/perlXPUtil.h, XPCOM/test.pl, XPCOM/typemap,
	XPCOM/GenericFactory/Changes,
	XPCOM/GenericFactory/GenericFactory.pm,
	XPCOM/GenericFactory/MANIFEST, XPCOM/GenericFactory/Makefile.PL,
	XPCOM/GenericFactory/README, XPCOM/GenericFactory/test.pl,
	XPCOM/GenericModule/Changes, XPCOM/GenericModule/GenericModule.pm,
	XPCOM/GenericModule/MANIFEST, XPCOM/GenericModule/Makefile.PL,
	XPCOM/GenericModule/test.pl, doc/docbook.css, doc/xpcom-perl.html,
	doc/xpcom-perl.xml, src/Makefile.in, src/Makefile.viper,
	src/perlXPUtil.cpp, src/perlXPUtil.h, src/plPerlBase.cpp,
	src/plPerlComponentLoader.cpp, src/plPerlFactory.cpp,
	src/plPerlModule.cpp, src/plPerlXPCOM_Base.h,
	src/plPerlXPModule.cpp, src/plPerlXPTCStub.cpp, src/xcDll.cpp,
	src/xcDll.h, tests/Makefile.in, tests/nsISample.idl,
	tests/nsSample.pm, tests/nsTestPerlSample.cpp: Checking in updated
	Perl XPCOM in its new location.



perlXPUtil.h
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is the Perl XPCOM language bindings.
 *
 * The Initial Developer of the Original Code is
 * Jumpline.com, Inc. Portions created by Jumpline.com, Inc. are 
 * Copyright (C) 2001 Jumpline.com, Inc. All Rights Reserved.
 *
 * Contributor(s):
 * Matt Kennedy <matt@jumpline.com> (original author)
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

/*****************************************************************************
 * File: perlXPUtil.h
 * Created by: matt
 * Created on: Wed 30 May 2001 03:14:23 PM EDT
 * $Id: perlXPUtil.h,v 1.1 2005/03/31 08:13:29 poumeyrol Exp $
 *****************************************************************************/
#ifndef PERLXPUTIL_H_
#define PERLXPUTIL_H_

#include "nsString.h"
#include "nsReadableUtils.h"
#include "nsCRT.h"
#include "nsCOMPtr.h"
#include "nsIServiceManager.h"
//#include "nsXPIDLString.h"
#include "nsISupports.h"
#include "nsIInterfaceInfo.h"
#include "nsIInterfaceInfoManager.h"
#include "nsMemory.h"
#include "xptinfo.h"
#include "xptcall.h"

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#define VP_RETURN_UNDEF \
  ST(0) = sv_newmortal(); \
  sv_setsv(ST(0), &PL_sv_undef); \
  XSRETURN(1);

#define VP_SET_RETCODE(ret_) \
{ \
  SV *_retsv_ = perl_get_sv("Components::returnCode", 1); \
  if(_retsv_) { \
    sv_setuv(_retsv_, (UV)ret_); \
  } \
}

#define VP_GET_RETCODE(ret_) \
{ \
  SV *_retsv_ = perl_get_sv("Components::returnCode", 1); \
  if(_retsv_) { \
    ret_ = (nsresult)SvUV(_retsv_); \
  } \
}

#define MAX_XPTC_ARG_COUNT 16

// Uncomment to turn on verbose bitching
//#define XP_PERL_DEBUG 1

// This is for making perl's typemapping happy, even though the typemaps are 
// never used...
typedef nsISupports rawISupports;

class perlXPUtil {
  public:
    static nsresult fetchXPObject(SV *theSV,
                                  nsCID **theCID,
                                  nsIID **theIID,
                                  nsISupports **theObject);

    static SV* makeXPObject(nsCID *cid, nsIID *iid, nsISupports *obj);
    
    static PRBool lookupConstant(nsIInterfaceInfo *iFace,
                                 const char *name,
                                 const nsXPTConstant **idx);

    static SV* constantToSV(const nsXPTConstant *theConstant, PRBool *result);

    static PRUint32 getSizeOfType(nsXPTType& theType);

    static void *convertSVToArray(SV *avRef,
                                  PRUint16 dimension,
                                  nsIInterfaceInfo *iface,
                                  nsXPTParamInfo *theParam,
                                  PRUint16 methodIndex,
                                  nsresult *res);

    static SV *convertArrayToSV(void *arrPtr,
                                PRUint16 dimension,
                                nsIInterfaceInfo *iface,
                                nsXPTParamInfo *theParam,
                                PRUint16 methodIndex,
                                nsresult *res);

    static nsresult fetchIIDFromSV(SV *theIID, nsIID& anIID);

    static nsresult GetInterfaceInfo(nsIInterfaceInfoManager *iim, 
                                 const nsIID& iid,
                                 nsIInterfaceInfo **iface);

    /**
      * Iterates through an array of interface ids to find one that supports 
      * the requested method.
      */
    static nsresult FindMethod(nsIInterfaceInfoManager *iim,
                           PRUint32 count,
                           nsIID **array,
                           const char *meth,
                           PRUint16 inputCount,
                           PRUint16 *methIndex,
                           const nsXPTMethodInfo **info,
                           nsIInterfaceInfo **retIFace);

    /**
      * Utility function that QI's an object for support for nsIClassInfo 
      * and if the object supports it fetches the supported interfaces via 
      * nsIClassInfo::GetInterfaces
      */
    static nsresult GetInterfaceList(nsISupports *object,
                                 PRUint32 *count,
                                 nsIID * **array);

    static nsresult GetMethodInfo(nsIInterfaceInfo *iface, 
                              const char *methName,
                              PRUint16 inputCount,
                              PRUint16 *methIndex,
                              const nsXPTMethodInfo **info);

    static void *ConvertSVToArray(SV *avRef, 
                          PRUint16 dimension,
                          nsIInterfaceInfo *iface,
                          nsXPTParamInfo *theParam,
                          PRUint16 methodIndex,
                          nsresult *res);

    static SV *ConvertArrayToSV(void *arrPtr,
                        PRUint16 dimension,
                        nsIInterfaceInfo *iface,
                        nsXPTParamInfo *theParam,
                        PRUint16 methodIndex,
                        nsresult *res);

    // Native to SV converters
    static nsresult ConvertInt8ToSV(PRInt8 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setiv(ret, (IV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertInt16ToSV(PRInt16 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setiv(ret, (IV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertInt32ToSV(PRInt32 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setiv(ret, (IV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertInt64ToSV(PRInt64 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setiv(ret, (IV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertUint8ToSV(PRUint8 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setuv(ret, (UV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertUint16ToSV(PRUint16 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setuv(ret, (UV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertUint32ToSV(PRUint32 in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setuv(ret, (UV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertUint64ToSV(PRUint64 in, SV **out) {
      *out = NULL;
      SV *ret = sv_newmortal();
      sv_setuv(ret, (UV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertFloatToSV(float in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setnv(ret, (double)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertDoubleToSV(double in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setnv(ret, (double)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertBoolToSV(PRBool in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      ret = boolSV(in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertCharToSV(char in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setpvn(ret, (char *)&in, 1);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertWCharToSV(PRUnichar in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      sv_setiv(ret, (IV)in);
      if(!ret) {
        return NS_ERROR_FAILURE;
      }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertIDToSV(nsID& in, SV **out) {
      SV *ret = sv_newmortal();
      *out = NULL;
      nsID *id = new nsID;
      if(!id) return NS_ERROR_OUT_OF_MEMORY;
      *id = in;
      sv_setref_pv(ret, "nsID", id);
      if(!ret) return NS_ERROR_FAILURE;
      *out = ret;
      return NS_OK;
    }

    // char * to SV (UTF8 or ISO)
    static nsresult ConvertStringToSV(const char *in, PRBool bUTF8, 
                                         SV **out) {
      SV *ret = sv_newmortal();
      if(!ret) return NS_ERROR_FAILURE;
      *out = NULL;
      sv_setpv(ret, in);
      if(bUTF8)
          SvUTF8_on(ret);
      *out = ret;
      return NS_OK;
    }

    // nsACString* to SV (UTF8 or ISO)
    static nsresult ConvertACStringToSV( nsACString* str, 
                                            PRBool bUTF8, SV **out) {
      return ConvertStringToSV(nsCString(*str).get(), PR_FALSE, out);
    }

    // PRUnichar* to SV (UTF8)
    static nsresult ConvertWStringToSV(const PRUnichar *in, SV **out) {
      return ConvertStringToSV(NS_ConvertUCS2toUTF8(in).get(), PR_TRUE, out);
    }

    // AString to SV (UTF8)
    static nsresult ConvertAStringToSV(nsAString* str, SV **out) {
      return ConvertWStringToSV(nsString(*str).get(), out);
    }

    static nsresult
    ConvertISupportsToSV(nsISupports *in, nsIID *iid, nsCID *cid, SV **out) {
      *out = perlXPUtil::makeXPObject(cid, iid, in);
      if(!*out) return NS_ERROR_FAILURE;
      return NS_OK;
    }

    // SV To Native converters
    static nsresult ConvertSVToInt8(SV *in, PRInt8 *out) {
      *out = (PRInt8)SvIV(in);
      return NS_OK;
    }

   static  nsresult ConvertSVToInt16(SV *in, PRInt16 *out) {
      *out = (PRInt16)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToInt32(SV *in, PRInt32 *out) {
      *out = (PRInt32)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToInt64(SV *in, PRInt64 *out) {
      *out = (PRInt64)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToUint8(SV *in, PRUint8 *out) {
      *out = (PRUint8)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToUint16(SV *in, PRUint16 *out) {
      *out = (PRUint16)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToUint32(SV *in, PRUint32 *out) {
      *out = (PRUint32)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToUint64(SV *in, PRUint64 *out) {
      *out = (PRUint64)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToFloat(SV *in, float *out) {
      *out = (float)SvNV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToDouble(SV *in, double *out) {
      *out = (double)SvNV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToBool(SV *in, PRBool *out) {
      *out = (PRBool)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToChar(SV *in, char *out) {
      *out = (char)*SvPV(in,PL_na);
      return NS_OK;
    }

    static nsresult ConvertSVToWChar(SV *in, PRUnichar *out) {
      *out = (PRUnichar)SvIV(in);
      return NS_OK;
    }

    static nsresult ConvertSVToID(SV *in, nsID *out) {
      if(SvROK(in)) { // Better checking needed
        nsID *id = (nsID *)SvIV((SV*)SvRV(in));
        *out = *id;
        return NS_OK;
      }
      return NS_ERROR_FAILURE;
    }

    static nsresult ConvertSVToString(SV *in, PRBool bUTF8, char **out) {
      if(SvUTF8(in) && !bUTF8)
      {
          SV* conv = sv_newmortal();
          sv_setsv(conv, in);
          sv_utf8_downgrade(conv, TRUE);
          in = conv;
      }
      if(!SvUTF8(in) && bUTF8)
      {
          SV* conv = sv_newmortal();
          sv_setsv(conv, in);
          sv_utf8_upgrade(conv);
          in = conv;
      }
      PRUint32 len;
      char *s = SvPV(in, len);
      char *ret = (char*) nsMemory::Clone(s, len);
      if(!ret) { return NS_ERROR_OUT_OF_MEMORY; }
      *out = ret;
      return NS_OK;
    }

    static nsresult ConvertSVToACString(SV* in, PRBool bUTF8, 
            nsACString& out)
    {
        char* pcTmp;
        nsresult res = ConvertSVToString(in, bUTF8, &pcTmp);
        if(NS_FAILED(res))
        {
            nsMemory::Free(pcTmp);
            return res;
        }
        out.Assign(pcTmp);
        return NS_OK;
    }

    static nsresult ConvertSVToWString(SV *in, PRUnichar **out) {
      if(SvUTF8(in))
        *out = ToNewUnicode(NS_ConvertUTF8toUCS2(SvPV_nolen(in)));
      else
        *out = ToNewUnicode(NS_ConvertASCIItoUCS2(SvPV_nolen(in)));
      if(!*out) {
        return NS_ERROR_OUT_OF_MEMORY;
      }
      return NS_OK;
    }

    static nsresult ConvertSVToAString(SV* in, nsAString& out)
    {
      if(SvUTF8(in))
        out.Assign(ToNewUnicode(NS_ConvertUTF8toUCS2(SvPV_nolen(in))));
      else
        out.Assign(ToNewUnicode(NS_ConvertASCIItoUCS2(SvPV_nolen(in))));
      return NS_OK;
    }

    static SV *MakeUndef() {
      SV *ret = sv_newmortal();
      sv_setsv(ret, &PL_sv_undef);
      return ret;
    }

};
#endif


